tests/src/quickgui/app/main.cpp

```diff
@@ -0,0 +1,97 @@
+/***************************************************************************
+  main.cpp
+  --------------------------------------
+  Date                 : Nov 2017
+  Copyright            : (C) 2017 by Peter Petrik
+  Email                : zilolv at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include <QCommandLineParser>
+#include <QQmlApplicationEngine>
+#include <QQmlComponent>
+#include <QQmlContext>
+#include <QQmlError>
+
+#include "qgis.h"
+#include "qgsapplication.h"
+#include "qgsproject.h"
+#include "qgslayertree.h"
+#include "qgsmessagelog.h"
+#include "qgsquickutils.h"
+#include "qgslogger.h"
+
+int main( int argc, char *argv[] )
+{
+  // 1) Initialize QGIS
+  QgsApplication app( argc, argv, true );
+
+  // Set up the QSettings environment must be done after qapp is created
+  QCoreApplication::setOrganizationName( "QGIS" );
+  QCoreApplication::setOrganizationDomain( "qgis.org" );
+  QCoreApplication::setApplicationName( "QgsQuick Test App" );
+  QCoreApplication::setApplicationVersion( Qgis::QGIS_VERSION );
+
+  QCommandLineParser parser;
+  parser.addVersionOption();
+  parser.process( app );
+
+  QgsApplication::init();
+  QgsApplication::initQgis();
+
+  // 2) Load QGIS Project
+  QString dataDir( TEST_DATA_DIR );  // defined in CMakeLists.txt
+  QString projectFile = dataDir + "/quickapp_project.qgs";
+  QgsDebugMsg( QStringLiteral( "project file:  %1" ).arg( projectFile ) );
+  QgsProject project;
+  bool res = project.read( projectFile );
+  Q_ASSERT( res );
+
+  QQmlEngine engine;
+  engine.addImportPath( QgsApplication::qmlImportPath() );
+  engine.rootContext()->setContextProperty( "__project", &project );
+  engine.rootContext()->setContextProperty( "__layers", QVariant::fromValue( project.layerTreeRoot()->layerOrder() ) );
```

Comment:
Is that something that can be accessed through the project?
I'm afraid of having something highly dynamic like the layer list as a context property.

Reply:
`layerTreeRoot()`, nor `layerOrder()` is INVOCABLE. So I would need to change both of those classes plus register another types in qml plugin, or create a function in qgsquickutils.cpp. Since this is just a test (in real project one would need either project chooser or some layer tree item), I am not sure if we want to make such changes in core libraries?

Reply:
ah right, it's a test again, forget it

Reply:
forgotten